### PR TITLE
Remove SDL2 dependencies: Phase 1

### PR DESCRIPTION
**Pull request by IntelOrca**

### CODE REVIEW

**Comment by AaronVanGeffen** [comment]
Just to clarify: is the goal to get rid of SDL altogether, or just separate it from the logic to allow for GUI-less builds? I initially thought the latter, but the `openrct2-ui` tree and its platform-specific window contexts make me think the former is the case.

**Comment by IntelOrca** [comment]
@AaronVanGeffen SDL2 will still be used, we are just trying to extract the logic so that we can build a stand alone CLI version for servers.

The UI context stuff has been designed with SDL2 to be somewhat interchangeable - just so that porting can be done to a platform that SDL2 does not support.

File: src/openrct2-cli/Cli.cpp [hunk 1]
```diff
@@ -0,0 +1,30 @@
+#pragma region Copyright (c) 2014-2016 OpenRCT2 Developers
```
Comment (marijnvdwerf) [inline src/openrct2-cli/Cli.cpp]: 2017

File: src/openrct2-ui/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+# Sources
+file(GLOB_RECURSE OPENRCT2_UI_SOURCES
+    "${CMAKE_CURRENT_LIST_DIR}/*.c"
+    "${CMAKE_CURRENT_LIST_DIR}/*.cpp"
+    "${CMAKE_CURRENT_LIST_DIR}/*.h"
+    "${CMAKE_CURRENT_LIST_DIR}/*.hpp")
+
+# Outputs
+set (PROJECT openrct2)
```
Comment (marijnvdwerf) [inline src/openrct2-ui/CMakeLists.txt]: why the space?

File: src/openrct2-ui/CursorData.cpp [hunk 3]
```diff
@@ -14,9 +14,10 @@
  *****************************************************************************/
 #pragma endregion
 
-#include "Cursors.h"
+#include <openrct2/interface/Cursors.h>
+#include "CursorRepository.h"
 
-namespace Cursors
+namespace OpenRCT2 { namespace Ui
```
Comment (marijnvdwerf) [inline src/openrct2-ui/CursorData.cpp]: Is this supposed to be on a single line?

File: src/openrct2-ui/UiContext.macOS.mm [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+				NSAlert *alert = [[[NSAlert alloc] init] autorelease];
+				[alert addButtonWithTitle:@"OK"];
+				alert.messageText = [NSString stringWithUTF8String:message.c_str()];
+				[alert runModal];
+			}
+        }
+
+        std::string ShowFileDialog(SDL_Window * window, const FileDialogDesc &desc) override
+        {
+			@autoreleasepool
```
Comment (marijnvdwerf) [inline src/openrct2-ui/UiContext.macOS.mm]: File uses tabs for indenting

File: src/openrct2/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,314 @@
...
+# Items below are not supported by ICC
+ADD_CHECK_C_COMPILER_FLAG(CMAKE_C_FLAGS C_WARN_REDUNDANT_DECLS -Wredundant-decls)
+ADD_CHECK_C_COMPILER_FLAG(CMAKE_C_FLAGS C_WARN_IGNORED_QUALIFIERS -Wignored-qualifiers)
+ADD_CHECK_CXX_COMPILER_FLAG(CMAKE_CXX_FLAGS CXX_WARN_REDUNDANT_DECLS -Wredundant-decls)
+ADD_CHECK_CXX_COMPILER_FLAG(CMAKE_CXX_FLAGS CXX_WARN_IGNORED_QUALIFIERS -Wignored-qualifiers)
+
+# -Wstrict-overflow is only active when -fstrict-overflow is enabled, but -fstrict-overflow
+# is enabled on -O2, -O3, -Os. This should help catch bugs locally before they reach Travis
+# As of 2a435bf -Wstrict-overflow=1 passes, but higher values do not.
+set(COMMON_COMPILE_OPTIONS "${COMMON_COMPILE_OPTIONS} -fstrict-overflow")
```
Comment (marijnvdwerf) [inline src/openrct2/CMakeLists.txt]: Shouldn't COMMON_COMPILE OPTIONS be added to the c/cxx flags?

**Comment by marijnvdwerf** [comment]
Trying to get Cmake to work on macOS:
```
[ 93%] Linking CXX shared library libopenrct2.dylib
Undefined symbols for architecture x86_64:
  "_EVP_DigestSignFinal", referenced from:
      NetworkKey::Sign(unsigned char const*, unsigned long, char**, unsigned long*) in NetworkKey.cpp.o
  "_EVP_DigestSignInit", referenced from:
      NetworkKey::Sign(unsigned char const*, unsigned long, char**, unsigned long*) in NetworkKey.cpp.o
  "_EVP_DigestVerifyFinal", referenced from:
      NetworkKey::Verify(unsigned char const*, unsigned long, char const*, unsigned long) in NetworkKey.cpp.o
  "_EVP_DigestVerifyInit", referenced from:
      NetworkKey::Verify(unsigned char const*, unsigned long, char const*, unsigned long) in NetworkKey.cpp.o
  "_EVP_PKEY_CTX_ctrl", referenced from:
      NetworkKey::Generate() in NetworkKey.cpp.o
  "_EVP_PKEY_CTX_free", referenced from:
      NetworkKey::~NetworkKey() in NetworkKey.cpp.o
  "_EVP_PKEY_CTX_new_id", referenced from:
      NetworkKey::NetworkKey() in NetworkKey.cpp.o
  "_EVP_PKEY_keygen", referenced from:
      NetworkKey::Generate() in NetworkKey.cpp.o
  "_EVP_PKEY_keygen_init", referenced from:
      NetworkKey::Generate() in NetworkKey.cpp.o
  "_iconv", referenced from:
      _win1252_to_utf8 in localisation.c.o
  "_iconv_close", referenced from:
      _win1252_to_utf8 in localisation.c.o
  "_iconv_open", referenced from:
      _win1252_to_utf8 in localisation.c.o
ld: symbol(s) not found for architecture x86_64
```

File: src/openrct2/CMakeLists.txt (hunk 5 above)
Comment (IntelOrca) [inline src/openrct2/CMakeLists.txt]: Maybe, this was copied from the current one, so if you can improve it - by all means.

File: src/openrct2-ui/CursorData.cpp (hunk 3 above)
Comment (IntelOrca) [inline src/openrct2-ui/CursorData.cpp]: Yes, I intend on changing it to `OpenRCT2::Ui` when we can use C++17.

**Comment by marijnvdwerf** [comment]
Also:
```
CMake Warning (dev):
  Policy CMP0042 is not set: MACOSX_RPATH is enabled by default.  Run "cmake
  --help-policy CMP0042" for policy details.  Use the cmake_policy command to
  set the policy and suppress this warning.

  MACOSX_RPATH is not specified for the following targets:

   libopenrct2

This warning is for project developers.  Use -Wno-dev to suppress it.
```

File: src/openrct2-ui/TextComposition.cpp [hunk 6]
```diff
@@ -111,9 +111,8 @@ void TextComposition::HandleMessage(const SDL_Event * e)
             // Map Keypad enter to regular enter.
             key = SDLK_RETURN;
         }
-
-        // _lastKeyPressed = e.key.keysym.sym;
-        // _keysPressed[e.key.keysym.scancode] = 1;
+		
+		GetContext()->GetUiContext()->SetKeysPressed(key, e->key.keysym.scancode);
```
Comment (IntelOrca) [inline src/openrct2-ui/TextComposition.cpp]: Tabs!

File: src/openrct2-ui/UiContext.cpp [hunk 7]
```diff
@@ -217,6 +217,12 @@ class UiContext final : public IUiContext
     {
         SDL_SetWindowGrab(_window, value ? SDL_TRUE : SDL_FALSE);
     }
+	
+	void SetKeysPressed(uint32 keysym, uint8 scancode) override
+	{
+			_lastKeyPressed = keysym;
+			_keysPressed[scancode] = 1;
+	}
```
Comment (IntelOrca) [inline src/openrct2-ui/UiContext.cpp]: Tabs!

File: src/openrct2/ui/UiContext.h [hunk 8]
```diff
@@ -119,6 +119,7 @@ namespace OpenRCT2
             virtual void                SetCursorTrap(bool value) abstract;
             virtual const uint8 *       GetKeysState() abstract;
             virtual const uint8 *       GetKeysPressed() abstract;
+			virtual void				SetKeysPressed(uint32 keysym, uint8 scancode) abstract;
```
Comment (IntelOrca) [inline src/openrct2/ui/UiContext.h]: Tabs!

File: src/openrct2-ui/TextComposition.cpp (hunk 6 above)
Comment (rwjuk) [inline src/openrct2-ui/TextComposition.cpp]: Already fixed in 961a4dd.

**Comment by duncanspumpkin** [comment]
@rwjuk there is a plugin for visual studio that warns you if you try to save a file with mixed tabs and spaces. It's very useful for this project.

**Comment by Gymnasiast** [comment]
@duncanspumpkin Richard uses macOS and Linux, so a VS plug-in is not an option.

**Comment by pizza2004** [comment]
Actually, Visual Studio just came out on macOS, but I wouldn't personally use it anyway.

**Comment by IntelOrca** [comment]
@pizza2004 Visual Studio for Mac is very much a different application entirely so you can't use any extensions for VS in VS for Mac.

**Comment by marijnvdwerf** [comment]
What about a git commit hook script?

**Comment by pizza2004** [comment]
@IntelOrca Obviously, I just thought it was an interesting thing people might like to know.

**Comment by Gymnasiast** [comment]
If I attempt to load a save using the native dialog, no saves are listed. There is also no option to select 'All files' (there used to be). Furthermore, if I click 'Cancel', it will display an error message and force me to select a file, giving me no option to cancel other than forcibly killing the OpenRCT2 process.

**Comment by marijnvdwerf** [comment]
Fails at the linking step for me
```
[361/361] Linking CXX executable openrct2
FAILED: openrct2 
: && /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++  -std=gnu++14  -Wno-error=deprecated-declarations -Wno-error=objc-method-access -fstrict-aliasing -Werror -Wundef -Wmissing-declarations -Winit-self -Wall -Wno-unknown-pragmas -Wno-unused-function -Wno-missing-braces  -Wno-comment -Wshadow  -Wmissing-declarations -Wnonnull -DDEBUG=0 -fPIC  -Wno-error=date-time -Wnull-dereference -Wredundant-decls -Wignored-qualifiers -Wstrict-overflow=1 -std=gnu++14 -Wl,-search_paths_first -Wl,-headerpad_max_install_names  CMakeFiles/openrct2.dir/src/openrct2-ui/CursorData.cpp.o CMakeFiles/openrct2.dir/src/openrct2-ui/CursorRepository.cpp.o CMakeFiles/openrct2.dir/src/openrct2-ui/TextComposition.cpp.o CMakeFiles/openrct2.dir/src/openrct2-ui/Ui.cpp.o CMakeFiles/openrct2.dir/src/openrct2-ui/UiContext.Linux.cpp.o CMakeFiles/openrct2.dir/src/openrct2-ui/UiContext.Win32.cpp.o CMakeFiles/openrct2.dir/src/openrct2-ui/UiContext.cpp.o CMakeFiles/openrct2.dir/src/openrct2-ui/audio/AudioChannel.cpp.o CMakeFiles/openrct2.dir/src/openrct2-ui/audio/AudioContext.cpp.o CMakeFiles/openrct2.dir/src/openrct2-ui/audio/AudioMixer.cpp.o CMakeFiles/openrct2.dir/src/openrct2-ui/audio/FileAudioSource.cpp.o CMakeFiles/openrct2.dir/src/openrct2-ui/audio/MemoryAudioSource.cpp.o CMakeFiles/openrct2.dir/src/openrct2-ui/drawing/engines/SoftwareDrawingEngine.cpp.o CMakeFiles/openrct2.dir/src/openrct2-ui/drawing/engines/opengl/CopyFramebufferShader.cpp.o CMakeFiles/openrct2.dir/src/openrct2-ui/drawing/engines/opengl/DrawImageShader.cpp.o CMakeFiles/openrct2.dir/src/openrct2-ui/drawing/engines/opengl/DrawLineShader.cpp.o CMakeFiles/openrct2.dir/src/openrct2-ui/drawing/engines/opengl/FillRectShader.cpp.o CMakeFiles/openrct2.dir/src/openrct2-ui/drawing/engines/opengl/OpenGLAPI.cpp.o CMakeFiles/openrct2.dir/src/openrct2-ui/drawing/engines/opengl/OpenGLDrawingEngine.cpp.o CMakeFiles/openrct2.dir/src/openrct2-ui/drawing/engines/opengl/OpenGLFramebuffer.cpp.o CMakeFiles/openrct2.dir/src/openrct2-ui/drawing/engines/opengl/OpenGLShaderProgram.cpp.o CMakeFiles/openrct2.dir/src/openrct2-ui/drawing/engines/opengl/SwapFramebuffer.cpp.o CMakeFiles/openrct2.dir/src/openrct2-ui/drawing/engines/opengl/TextureCache.cpp.o CMakeFiles/openrct2.dir/src/openrct2-ui/UiContext.macOS.mm.o  -o openrct2  libopenrct2.dylib -lSDL2 -lSDL2_ttf -lSDL2 -framework OpenGL -ljansson -lspeexdsp -lpng16 -lz -lzip -lz -lzip -ldl -lcurl /usr/local/Cellar/openssl/1.0.2k/lib/libssl.dylib /usr/local/Cellar/openssl/1.0.2k/lib/libcrypto.dylib -lSDL2 -lSDL2_ttf -lSDL2 -lSDL2_ttf /usr/lib/libiconv.dylib && :
Undefined symbols for architecture x86_64:
  "_OBJC_CLASS_$_NSAlert", referenced from:
      objc-class-ref in UiContext.macOS.mm.o
  "_OBJC_CLASS_$_NSMutableArray", referenced from:
      objc-class-ref in UiContext.macOS.mm.o
  "_OBJC_CLASS_$_NSOpenPanel", referenced from:
      objc-class-ref in UiContext.macOS.mm.o
  "_OBJC_CLASS_$_NSSavePanel", referenced from:
      objc-class-ref in UiContext.macOS.mm.o
  "_OBJC_CLASS_$_NSString", referenced from:
      objc-class-ref in UiContext.macOS.mm.o
  "_OBJC_CLASS_$_NSURL", referenced from:
      objc-class-ref in UiContext.macOS.mm.o
  "_OBJC_CLASS_$_NSWindow", referenced from:
      objc-class-ref in UiContext.macOS.mm.o
  "___CFConstantStringClassReference", referenced from:
      CFString in UiContext.macOS.mm.o
      CFString in UiContext.macOS.mm.o
      CFString in UiContext.macOS.mm.o
      CFString in UiContext.macOS.mm.o
      CFString in UiContext.macOS.mm.o
  "_objc_autoreleasePoolPop", referenced from:
      OpenRCT2::Ui::macOSContext::macOSContext() in UiContext.macOS.mm.o
      OpenRCT2::Ui::macOSContext::ShowMessageBox(SDL_Window*, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) in UiContext.macOS.mm.o
      OpenRCT2::Ui::macOSContext::ShowFileDialog(SDL_Window*, OpenRCT2::Ui::FileDialogDesc const&) in UiContext.macOS.mm.o
      OpenRCT2::Ui::macOSContext::ShowDirectoryDialog(SDL_Window*, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) in UiContext.macOS.mm.o
  "_objc_autoreleasePoolPush", referenced from:
      OpenRCT2::Ui::macOSContext::macOSContext() in UiContext.macOS.mm.o
      OpenRCT2::Ui::macOSContext::ShowMessageBox(SDL_Window*, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) in UiContext.macOS.mm.o
      OpenRCT2::Ui::macOSContext::ShowFileDialog(SDL_Window*, OpenRCT2::Ui::FileDialogDesc const&) in UiContext.macOS.mm.o
      OpenRCT2::Ui::macOSContext::ShowDirectoryDialog(SDL_Window*, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) in UiContext.macOS.mm.o
  "_objc_msgSend", referenced from:
      OpenRCT2::Ui::macOSContext::macOSContext() in UiContext.macOS.mm.o
      OpenRCT2::Ui::macOSContext::ShowMessageBox(SDL_Window*, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) in UiContext.macOS.mm.o
      OpenRCT2::Ui::macOSContext::ShowFileDialog(SDL_Window*, OpenRCT2::Ui::FileDialogDesc const&) in UiContext.macOS.mm.o
      OpenRCT2::Ui::macOSContext::ShowDirectoryDialog(SDL_Window*, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) in UiContext.macOS.mm.o
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
ninja: build stopped: subcommand failed.
```

**Comment by rwjuk** [comment]
@Gymnasiast What OS was this on?

**Comment by rwjuk** [comment]
Tested and working on macOS 10.12

- Native file load/save dialogs
- TTF display (tested Japanese)
- Pointing game to RCT2 & RCT1 path on initial setup (The bug that #5500 fixed is still there, but a rebase will cure that. Setting RCT1 path works apart from that.)

**Comment by Gymnasiast** [comment]
@rwjuk It was Ubuntu 17.04. I amended my comment.

**Comment by Gymnasiast** [comment]
Tested it.

Zenity and Kdialog now appear to work correctly, but the 'All files' item is still missing.

**Comment by Gymnasiast** [comment]
Ok, the All files filter works, but attempting to load a file with another extension does not. For example,. rename one your saves from mypark.sv6 to mypark.sv6.bak and it won't load.

**Comment by IntelOrca** [comment]
@Gymnasiast this happens on develop too, so its not a regression from this PR.

**Comment by Gymnasiast** [comment]
Hm, I see. In that case, include the comment and it's fine to merge by me.

File: src/openrct2-cli/Cli.cpp [hunk 9]
```diff
@@ -1,4 +1,4 @@
-#pragma region Copyright (c) 2014-2016 OpenRCT2 Developers
+#pragma region Copyright (c) 2014-2017 OpenRCT2 Developers
```
Comment (rwjuk) [inline src/openrct2-cli/Cli.cpp]: Should we change this on all files that've been modified?

Comment (janisozaur) [inline src/openrct2-cli/Cli.cpp]: whenever we remember

Comment (Gymnasiast) [inline src/openrct2-cli/Cli.cpp]: It would be neater, but I'm not sure how long we should be holding this PR back.

Comment (rwjuk) [inline src/openrct2-cli/Cli.cpp]: Oh I'm absolutely not advocating holding this back for trivia like copyright years - more of a general question. We can go through and update all files that've been touched this year in one sweep some other time.

Comment (Gymnasiast) [inline src/openrct2-cli/Cli.cpp]: @rwjuk Sounds good, could you create an issue for that and put it on the milestone?

Comment (rwjuk) [inline src/openrct2-cli/Cli.cpp]: @Gymnasiast Done #5516

**Comment by janisozaur** [comment]
@grahamedgecombe if you're around, can you please also take a look at this? We want to have `libopenrct2.so` and its clients: `openrct2` (GUI) and `openrct2-cli` (headless build), how would we best address the lookup of `libopenrct2.so` for portable builds? It may also affect packaging, which I haven't inspected.

**Comment by grahamedgecombe** [comment]
@janisozaur for portable builds set an rpath with `$ORIGIN` maybe?

**Comment by Gymnasiast** [comment]
With this last commit it appears to work correctly on my system. Moving both the binary and lib from the build dir to the executable dir now works as intended.

File: CMakeLists.txt [hunk 10]
```diff
@@ -81,11 +82,15 @@ install(CODE "file(DOWNLOAD ${TITLE_SEQUENCE_URL} \$ENV{DESTDIR}${CMAKE_INSTALL_
...
-else()
-    install(TARGETS "libopenrct2" LIBRARY DESTINATION lib)
-endif()
-install(TARGETS "openrct2" RUNTIME DESTINATION bin)
+    install(TARGETS "libopenrct2" RUNTIME DESTINATION "bin")
+else ()
+    if (PORTABLE)
+        install(TARGETS "libopenrct2" LIBRARY DESTINATION "bin")
+    else ()
+        install(TARGETS "libopenrct2" LIBRARY DESTINATION "usr/local/lib")
```
Comment (grahamedgecombe) [inline CMakeLists.txt]: The DESTINATION here should just be "lib". cmake already adds the prefix, which may not be `/usr/local` - e.g. it's `/usr` in the Arch package.

Currently this results in `libopenrct2.so` being installed in `/usr/usr/local/lib`.

Comment (IntelOrca) [inline CMakeLists.txt]: So you think it should be `lib` rather than `local/lib`?

Comment (grahamedgecombe) [inline CMakeLists.txt]: Yeah, I think it should be `install(TARGETS "libopenrct2" LIBRARY DESTINATION "lib")`.

cmake adds the `/usr` or `/usr/local` (or whaever) prefix for you - based on the value of `CMAKE_INSTALL_PREFIX`.

Comment (IntelOrca) [inline CMakeLists.txt]: Thanks, I have changed it back to `lib`.

File: src/openrct2-ui/CMakeLists.txt [hunk 11]
```diff
@@ -75,6 +75,11 @@ if (WIN32)
     # tell it that it is
     set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -D__USE_MINGW_ANSI_STDIO=1")
 endif ()
+
+if (PORTABLE)
+    set(CMAKE_INSTALL_RPATH "$ORIGIN")
```
Comment (marijnvdwerf) [inline src/openrct2-ui/CMakeLists.txt]: Doesn't this get interpreted as "fill with contents of variable named 'ORIGIN'?

Comment (IntelOrca) [inline src/openrct2-ui/CMakeLists.txt]: CMAKE uses `${...}` for variables and `$ENV{...}` for environment variables. I think this is saw on various internet pages for how to do it in cmake.

Comment (rwjuk) [inline src/openrct2-ui/CMakeLists.txt]: `set(CMAKE_INSTALL_RPATH "$ORIGIN")` needs to go before `add_executable()` for it to work I think.

Comment (rwjuk) [inline src/openrct2-ui/CMakeLists.txt]: `set(CMAKE_INSTALL_RPATH "$ORIGIN")` needs to go before `add_executable()` I think.

Comment (janisozaur) [inline src/openrct2-ui/CMakeLists.txt]: Shouldn't the other frontend also receive the same treatment?

Comment (rwjuk) [inline src/openrct2-ui/CMakeLists.txt]: @janisozaur Derp. Yes, `set(CMAKE_INSTALL_RPATH "$ORIGIN")` added for the CLI build.

**Pull request merged**